src/ParallelAlgorithms/Amr/Actions/UpdateAmrDecision.hpp

```diff
@@ -0,0 +1,129 @@
+///
+/// DataBox:
+/// - Uses:
+///   * domain::Tags::Element<volume_dim>
+/// - Modifies:
+///   * amr::domain::Tags::NeighborFlags
+///   * amr::domain::Tags::Flags (if AMR decision is updated)
+///
+/// Invokes:
+/// - amr::Actions::UpdateAmrDecision on all neighboring Element%s (if AMR
+///   decision is updated)
+///
+/// \details This Element calls amr::domain::update_amr_decision to see if its
+/// AMR decision needs to be updated.  If it does, the Element will call
+/// amr::Actions::UpdateAmrDecision on its neighbors.
+///
+struct UpdateAmrDecision {
+  template <typename ParallelComponent, typename DbTagList,
+            typename Metavariables, typename ArrayIndex>
+  static void apply(
+      db::DataBox<DbTagList>& box, Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/,
+      const ElementId<Metavariables::volume_dim>& neighbor_id,
+      const std::array<amr::domain::Flag, Metavariables::volume_dim>&
+          neighbor_amr_flags) {
+    constexpr size_t volume_dim = Metavariables::volume_dim;
+
+    // Actions can be executed in any order.  Therefore we need to check:
+    // - If we have evaluated our own AMR decision.  If not, just store
+    //   the neighbor flags and return.
+    std::array<amr::domain::Flag, Metavariables::volume_dim>
+        my_current_amr_flags = get<amr::domain::Tags::Flags<volume_dim>>(box);
+    if (amr::domain::Flag::Undefined == my_current_amr_flags[0]) {
+      db::mutate<amr::domain::Tags::NeighborFlags<volume_dim>>(
+          make_not_null(&box),
+          [&neighbor_id,
+           &neighbor_amr_flags](const gsl::not_null<std::unordered_map<
+                                    ElementId<volume_dim>,
+                                    std::array<amr::domain::Flag, volume_dim>>*>
+                                    amr_flags_of_neighbors) {
+            auto& my_neighbor_amr_flags =
+                (*amr_flags_of_neighbors)[neighbor_id];
+            for (size_t d = 0; d < volume_dim; ++d) {
+              my_neighbor_amr_flags[d] = neighbor_amr_flags[d];
+            }
+          });
+      return;
+    }
+    // - If the neighbor flags we just received have higher priority than
+    //   the flags we already received.  If not, then return.
+    const auto& my_neighbors_amr_flags =
+        get<amr::domain::Tags::NeighborFlags<volume_dim>>(box);
+    if (1 == my_neighbors_amr_flags.count(neighbor_id)) {
+      bool received_old_decision = true;
+      const auto& stored_neighbor_amr_flags =
+          my_neighbors_amr_flags.at(neighbor_id);
+      for (size_t d = 0; d < volume_dim; ++d) {
+        if (neighbor_amr_flags[d] > stored_neighbor_amr_flags[d]) {
+          received_old_decision = false;
+          break;
+        }
+      }
+      if (received_old_decision) {
+        return;
+      }
+    }
+
+    const auto& element = get<::domain::Tags::Element<volume_dim>>(box);
+    const bool my_amr_decision_changed = amr::domain::update_amr_decision(
```

Comment:
A comment here would be useful. The return value isn't used until later, but presumably the `not_null` change is needed for the `mutate`

Reply:
realized I could simplify this by using `db::get_mutable_reference`.  I think it improves the flow...